common/client-libs/gateway-client/src/bandwidth.rs

```diff
@@ -0,0 +1,190 @@
+        json::parse(ETH_JSON_ABI)
+            .expect("Invalid json abi")
+            .dump()
+            .as_bytes(),
+    )
+    .expect("Invalid json abi")
+}
+
+#[derive(Clone)]
+pub struct BandwidthController {
+    #[cfg(feature = "coconut")]
+    validator_endpoints: Vec<url::Url>,
+    #[cfg(feature = "coconut")]
+    identity: PublicKey,
+    #[cfg(not(feature = "coconut"))]
+    contract: Contract<Http>,
+    #[cfg(not(feature = "coconut"))]
+    eth_private_key: SecretKey,
+}
+
+impl BandwidthController {
+    #[cfg(feature = "coconut")]
+    pub fn new(validator_endpoints: Vec<url::Url>, identity: PublicKey) -> Self {
+        BandwidthController {
+            validator_endpoints,
+            identity,
+        }
+    }
+
+    #[cfg(not(feature = "coconut"))]
+    pub fn new(eth_endpoint: String, eth_private_key: String) -> Result<Self, GatewayClientError> {
+        // Fail early, on invalid url
+        let transport =
+            Http::new(&eth_endpoint).map_err(|_| GatewayClientError::InvalidURL(eth_endpoint))?;
+        let web3 = web3::Web3::new(transport);
+        // Fail early, on invalid abi
+        let contract = eth_contract(web3);
+        let eth_private_key = secp256k1::SecretKey::from_str(&eth_private_key)
+            .map_err(|_| GatewayClientError::InvalidEthereumPrivateKey)?;
+
+        Ok(BandwidthController {
+            contract,
+            eth_private_key,
+        })
+    }
+
+    #[cfg(feature = "coconut")]
+    pub async fn prepare_coconut_credential(
+        &self,
+    ) -> Result<coconut_interface::Credential, GatewayClientError> {
+        let verification_key = obtain_aggregate_verification_key(&self.validator_endpoints).await?;
+
+        let bandwidth_credential =
+            obtain_signature(&self.identity.to_bytes(), &self.validator_endpoints).await?;
+        // the above would presumably be loaded from a file
+
+        // the below would only be executed once we know where we want to spend it (i.e. which gateway and stuff)
+        Ok(prepare_for_spending(
+            &self.identity.to_bytes(),
+            &bandwidth_credential,
+            &verification_key,
+        )?)
+    }
+
+    #[cfg(not(feature = "coconut"))]
+    pub async fn prepare_token_credential(
+        &self,
+        gateway_identity: PublicKey,
+    ) -> Result<TokenCredential, GatewayClientError> {
+        let mut rng = OsRng;
+
+        let kp = identity::KeyPair::new(&mut rng);
+
+        let verification_key = *kp.public_key();
+        let signed_verification_key = kp.private_key().sign(&verification_key.to_bytes());
+        self.buy_token_credential(verification_key, signed_verification_key)
+            .await?;
+
+        let message: Vec<u8> = verification_key
+            .to_bytes()
+            .iter()
+            .chain(gateway_identity.to_bytes().iter())
+            .copied()
+            .collect();
+        let signature = kp.private_key().sign(&message);
+
+        Ok(TokenCredential::new(
```

Comment:
Yes, this is the next step I'm working on, storing the private key somewhere such that, in case of your scenario (which I think is on some error path probably), there is still a way to recover from it. I'll do that in a separate PR though.